### PR TITLE
Fixes #23523 - introduced optional search for used object on all_hosts

### PR DESCRIPTION
**Pull request by timdeluxe**
See this post https://projects.theforeman.org/issues/23523?issue_count=10&issue_position=5&next_issue_id=21046&prev_issue_id=23524#note-13 for details (or alternatively the changes to the README.md in the commit

### CODE REVIEW

**Comment by lzap** [comment]
Thank you for this very nice patch, makes sense.

I no longer have a working Infoblox setup, so let me ping @ares to find somebody to get this tested.

**Comment by timdeluxe** [comment]
Thanks for the flowers.
After using the modified version by myself i noticed, that the view in the Foreman UI on the "DHCP" button under the affected subnet in "Subnets" also shows those "dummy" items. If that is an advantage or disadvantage, everybody can decide for themselves.
Only alternative would reimplement the "unused_ip" method in the plugin, along with its own version of "all_hosts" (and maybe also an alternative version of "build_reservation"). I don't know if that would be useful or just raises complexity unnecessarily, i am open for ideas...

**Comment by lzap** [comment]
Dummy items?

Maybe @ekohl could drop his view on this?

**Comment by timdeluxe** [comment]
> Dummy items?

It is described in redmine post already, however in short: So far (before this change) only Infoblox "Host"s (if record_type = host) plus leases or "Fixedadress"s plus leases show up in the Foreman UI > Subnets > affected Subnet > DHCP view (and most probably on the Foreman REST API as well, did not check that). With my change also other objects like a DNS A entry show up in that list as "DHCP" reservation, despite the fact, that those are not DHCP reservations.

This is only for the "all_hosts" view, all others are untouched.

Only solution to that would be the approach i describe in https://github.com/theforeman/smart_proxy_dhcp_infoblox/pull/50#issuecomment-1241618031

File: lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb
Comment (ekohl) [inline lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_plugin.rb:22]: Instead of creating a custom validator, you can also use the Enum validator (introduced in https://github.com/theforeman/smart-proxy/commit/685c026e30ade098f1748fb8c5a0dd16e993efec)
```suggestion
    validate :used_ips_search_type, enum: %w[record_type used]
```

If you do so, please raise the minimum version on line 14/15 (`requires :dhcp`) to `>= 3.2`.

**Review by ekohl — COMMENTED** [review]
I'll need to do some further reading on your research here since I'm not familiar with Infoblox.

There is one thing that did jump out to me was the plugin validator. I see you got there by copying the existing implementation, but in https://github.com/theforeman/smart_proxy_dhcp_infoblox/pull/51 I fixed that.

Another thing is that CI doesn't appear to be enabled on this repo, which I'll also need to get into.

**Comment by timdeluxe** [comment]
> I'll need to do some further reading on your research here since I'm not familiar with Infoblox.

If you need any information, let me know.

> There is one thing that did jump out to me was the plugin validator. I see you got there by copying the existing implementation, but in #51 I fixed that.

Thanks, i fixed that now and raised the dhcp version as proposed. I fixed only my new validator, since the record_type validator will be fixed by your PR. (Whenever it gets merged)

> Another thing is that CI doesn't appear to be enabled on this repo, which I'll also need to get into.

Can't help much with that, but yes, i noticed that as well. I did run the unit- and rubocop tests and all is green, but it is better to see it with CI instead of believing me.

**Comment by ekohl** [comment]
> Thanks, i fixed that now and raised the dhcp version as proposed. I fixed only my new validator, since the record_type validator will be fixed by your PR. (Whenever it gets merged)

I just merged it and other changes, but those introduced merged conflicts. Would you mind rebasing?

> Can't help much with that, but yes, i noticed that as well. I did run the unit- and rubocop tests and all is green, but it is better to see it with CI instead of believing me.

It is because Jenkins doesn't recognize you as a user, so first someone has to approve the workflow.

File: lib/smart_proxy_dhcp_infoblox/fixed_address_crud.rb
Comment (ekohl) [inline lib/smart_proxy_dhcp_infoblox/fixed_address_crud.rb:12]: Rather than declaring this again, I think you can rely on the parent class (you do call `super()` correctly).

File: lib/smart_proxy_dhcp_infoblox/host_ipv4_address_crud.rb
Comment (ekohl) [inline lib/smart_proxy_dhcp_infoblox/host_ipv4_address_crud.rb:12]: Same here: you can probably rely on the parent class.

**Review by ekohl — COMMENTED** [review]
To me it reads well. I don't have an Infoblox setup to verify this, but given it doesn't change the defaults nor the default implementation I think it can be merged. Just 2 small details inline.

File: config/dhcp_infoblox.yml.example
Comment (dploeger) [inline config/dhcp_infoblox.yml.example:16]:
```suggestion
# Search type used when looking for used IPs: can be "record_type" or "used". The latter can be used when you want to include all "USED"
```

**Review by dploeger — CHANGES_REQUESTED** [review]
Just a small doc fix

**Comment by dploeger** [comment]
> Besides that stuff: What about the logic i described here [#50 (comment)](https://github.com/theforeman/smart_proxy_dhcp_infoblox/pull/50#issuecomment-1241618031) - is it worth the work, before we merge? I don't want to be the only one to decide.

I'd vote against implementing it, because it only adds complexity and the users using this plugin and have set the option to "used" *know* that it's not only DHCP because they actively decided against it.

**Comment by ekohl** [comment]
ok to test

**Review by ekohl — APPROVED** [review]
> > > Can't help much with that, but yes, i noticed that as well. I did run the unit- and rubocop tests and all is green, but it is better to see it with CI instead of believing me.
> > 
> > 
> > It is because Jenkins doesn't recognize you as a user, so first someone has to approve the workflow.
> 
> Well, then i hope "someone" can do it?

Done. Just wanted to wait last time since there was a merge conflict.

> > To me it reads well. I don't have an Infoblox setup to verify this, but given it doesn't change the defaults nor the default implementation I think it can be merged. Just 2 small details inline.
> 
> Thanks for pointing them out. Corrected those and a bunch of other rubocop offenses, which were first raised now after your rubocop updates.
> 
> Besides that stuff: What about the logic i described here [#50 (comment)](https://github.com/theforeman/smart_proxy_dhcp_infoblox/pull/50#issuecomment-1241618031) - is it worth the work, before we merge? I don't want to be the only one to decide.

Like @dploeger I think it's not worth it and go with the current implementation.

@dploeger @timdeluxe this (and `smart_proxy_dns_infoblox`) could use more maintainers. Would you be interested?

**Comment by timdeluxe** [comment]
> Done. Just wanted to wait last time since there was a merge conflict.

Thanks

> > Besides that stuff: What about the logic i described here [#50 (comment)](https://github.com/theforeman/smart_proxy_dhcp_infoblox/pull/50#issuecomment-1241618031) - is it worth the work, before we merge? I don't want to be the only one to decide.
> 
> Like @dploeger I think it's not worth it and go with the current implementation.

Okay, there is still the chance to do it later in another PR. How about a release now since it is merged? (Or should we do that on our own once we became maintainers?)

> @dploeger @timdeluxe this (and `smart_proxy_dns_infoblox`) could use more maintainers. Would you be interested?

Yes, at least i am. We have a running Infoblox setup, so we can test changes against the real API. Maybe some time in the future we also need to adopt/fork the repo of the "infoblox" gem, since its owner archived it. Right now it still works and it fits to the current API spec, but it might change in the future.

What do i (or we) have to do to become maintainers?

**Comment by ekohl** [comment]
> Yes, at least i am. We have a running Infoblox setup, so we can test changes against the real API.

Excellent

> Maybe some time in the future we also need to adopt/fork the repo of the "infoblox" gem, since its owner archived it. Right now it still works and it fits to the current API spec, but it might change in the future.

Yes, that's also a concern I have. Might be even nicer if we can convince the author to transfer the repository.

> What do i (or we) have to do to become maintainers?

I've just invited you to the theforeman org on GH and add you to the Infoblox team. I also added the title Infoblox to you on Discourse.

**Comment by timdeluxe** [comment]
@ekohl Can you point me to docs, which describe how to release the plugin and kick off the (debian) packaging? Or is a change like this https://github.com/theforeman/smart_proxy_dhcp_infoblox/commit/80bf4371b4c01d25dbea7cf093badcc9ed858229 plus a new version git tag enough to have the automation (Jenkins) do it?

**Comment by timdeluxe** [comment]
> @ekohl Can you point me to docs, which describe how to release the plugin and kick off the (debian) packaging? Or is a change like this [80bf437](https://github.com/theforeman/smart_proxy_dhcp_infoblox/commit/80bf4371b4c01d25dbea7cf093badcc9ed858229) plus a new version git tag enough to have the automation (Jenkins) do it?

@lzap Maybe you can help me as with this specific question as well?

**Comment by ekohl** [comment]
@timdeluxe today we don't have automation for that. I've been meaning to get to it (https://community.theforeman.org/t/use-github-actions-to-release-gems/24583) but we don't have it yet.

So today the process is:
* Bump version in `lib/smart_proxy_dhcp_infoblox/dhcp_infoblox_version.rb`
* Commit and tag as `vX.Y.Z`
* `gem build smart_proxy_dhcp_infoblox.gemspec && gem push smart_proxy_dhcp_infoblox-X.Y.Z.gem`

Once the gem is out, we can update packaging. We have weekly automation (https://github.com/theforeman/foreman-packaging/blob/master/.github/workflows/bump_packages.yml) but you can also submit it manually if you don't want to wait.

I'd also be happy to write the GitHub Action to release in this repo and use it as the first repo to get it rolling.

**Comment by timdeluxe** [comment]
@ekohl Ah thanks for feedback.

> I'd also be happy to write the GitHub Action to release in this repo and use it as the first repo to get it rolling.

I leave the decision up to you. Please let me know how you decide.

I registered now on rubygems with the same username like here on github (timdeluxe). I think in order for me to be able to push to rubygems, one of you need to add me as maintainer to the smart_proxy_dhcp_infoblox (and mabye smart_proxy_dns_infoblox) as well. Even if you implement the GitHub Action this may be wise?

> Once the gem is out, we can update packaging. We have weekly automation (https://github.com/theforeman/foreman-packaging/blob/master/.github/workflows/bump_packages.yml) but you can also submit it manually if you don't want to wait.

Weekly automation is fine for me so far. Also i can't submit the workflow yet due to lack of permissions.

Let me know your decision and if you decide for writing the GitHub Action when it is done. Depending on that i will either do only step 1 and 2 or all three steps of your process description.

**Comment by ekohl** [comment]
> I leave the decision up to you. Please let me know how you decide.

I just merged https://github.com/theforeman/smart_proxy_dhcp_infoblox/pull/54 so please give it a go.

> I registered now on rubygems with the same username like here on github (timdeluxe). I think in order for me to be able to push to rubygems, one of you need to add me as maintainer to the smart_proxy_dhcp_infoblox (and mabye smart_proxy_dns_infoblox) as well. Even if you implement the GitHub Action this may be wise?

Rubygems ownership is a bit tedious, but it is historically how we've done it so I'd be ok with that.

> Weekly automation is fine for me so far. Also i can't submit the workflow yet due to lack of permissions.

Yes, the whole run is a bit verbose so you'd probably have call the update manual.

> Let me know your decision and if you decide for writing the GitHub Action when it is done. Depending on that i will either do only step 1 and 2 or all three steps of your process description.

Please try step 1 and 2, see how it goes.

**Comment by timdeluxe** [comment]
Looks good: https://rubygems.org/gems/smart_proxy_dhcp_infoblox/versions/0.0.17 👍

Now either wait some days or you call the packaging manually or i get permission to do so (or i am too stupid).

**Comment by ekohl** [comment]
> Now either wait some days or you call the packaging manually or i get permission to do so (or i am too stupid).

I don't think we can give permission just to call that action - you have to have full permissions on the packaging repo. Of course you can follow the manual steps: https://github.com/theforeman/foreman-packaging/tree/deb/develop#quick-update (which is what the action essentially does anyway).

**Comment by timdeluxe** [comment]
Okay, got it. I think i will just wait...

**Comment by ekohl** [comment]
> Sorry to bother you again here

No worries.

> Do i have to raise PRs now to cherry-pick the commit into the deb/3.3 and deb/3.4 branches?

Correct - cherry picks into stable branches are always manual. Mostly because finding out what's safe and what isn't is a hard problem.

**Pull request merged**
